### PR TITLE
Re-Optimised Dockerfile

### PR DESCRIPTION
**Pull request by telegrambotdev**
Re-Optimised Dockerfile with prebuilt ffmpeg-alpine image.Everything should work fine

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,27 @@
...
-ADD ./ /
+WORKDIR /app
 
-RUN apt-get -y update
-RUN apt-get -y upgrade
-RUN apt-get -y install git
-RUN apt-get -y install ffmpeg
-RUN pip install -r requirements.txt
+ENV DEBIAN_FRONTEND=noninteractive
+ENV TZ=Asia/Kolkata
```
Comment (retrodaredevil) [inline Dockerfile:6]: Maybe before this PR is merged, this line should be removed? I don't think defaulting the timezone to Asia/Kolkata is a good idea. It's probably best to leave the timezone as default or whatever the user specifies.

**Review by telegrambotdev — COMMENTED** [review]
removed timezone

**Comment by BassThatHertz** [comment]
@retrodaredevil or anyone else who is reading this, can you re-clone this repository as I've just pushed a commit, follow the instructions under "Building with Docker" in the README, and then try the YouTube downloader? I'm getting the following error when doing so:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/yt.py", line 262, in yt_downloader
    return send_json_response(session['progress_file_path'], video_id, ' [audio].')
  File "/app/yt.py", line 144, in send_json_response
    correct_file = [filename for filename in downloads if video_id in filename and download_type in filename][0]
IndexError: list index out of range
```
The most informative lines being:
```
  File "/app/yt.py", line 144, in send_json_response
    correct_file = [filename for filename in downloads if video_id in filename and download_type in filename][0]
IndexError: list index out of range
```
I'm wondering if anyone else will get the same issue. @telegrambotdev are you sure that you don't get the aforementioned error?

**Comment by retrodaredevil** [comment]
I haven't tested this in a while, but we could start using just youtube-dl as a dependency instead of `git+https://github.com/ytdl-org/youtube-dl.git@refs/pull/25717/head` like we currently have it. I think the change we needed a while back got merged (https://github.com/ytdl-org/youtube-dl/pull/25717/).

I'll try recloning and building the docker image if I find time.

**Comment by BassThatHertz** [comment]
> I haven't tested this in a while, but we could start using just youtube-dl as a dependency instead of `git+https://github.com/ytdl-org/youtube-dl.git@refs/pull/25717/head` like we currently have it. I think the change we needed a while back got merged ([ytdl-org/youtube-dl#25717](https://github.com/ytdl-org/youtube-dl/pull/25717)).

I thought I had already removed it from requirements.txt. Just had a look and it hasn't been removed, so I'll do so now and edit the README to list it as a dependency when running locally.

I would merge this pull request to make testing the new Dockerfile easier, but this will close this issue and I'd rather keep this issue open to make my above request more visible to people who view this repository.

**Comment by BassThatHertz** [comment]
@telegrambotdev @retrodaredevil Here's the reason for the issue I mentioned as well as the solution:

I explored inside the Docker container using `docker exec -it name-of-container bash`...

What I found is that youtube-dl isn't able to run with a simple `/usr/local/bin/youtube-dl`, I got the following error:
`env: can't execute 'python': No such file or directory`

The path to Python 3 also needs to be specified. So, to run youtube-dl in the container, you need to do something like `/usr/bin/python3 /usr/local/bin/youtube-dl`.

This means that yt.py needs to be edited before building the container. The way we are using youtube-dl is by passing a list to subprocess.run(), and therefore you need to make `youtube_dl_path` a list like so `youtube_dl_path = ['/usr/bin/python3', '/usr/local/bin/youtube-dl']`, and then concatenate all of the `args` lists in yt.py with this list. Here's an example:
```
        args = youtube_dl_path + ['--newline', '--restrict-filenames', '--cookies', 'cookies.txt',
                                  '-o', download_template, '--', video_id]
```
You can see the changes that needed to be made to yt.py in [this](https://github.com/BassThatHertz/AudioAndVideoConverter/commit/46448b9ec418c6d86512f86bceeb293892cc3635) commit.

**Comment by retrodaredevil** [comment]
Is there a reason you are calling youtube-dl as a subprocess? youtube-dl is written in Python and I think you are able to do the same thing by just calling functions from its public API. Then you don't have to worry about the path to the python interpreter or the path to the executable.

**Comment by BassThatHertz** [comment]
@retrodaredevil That was a good suggestion, so I've changed the code to use the API instead of subprocess.run(). https://github.com/BassThatHertz/AudioAndVideoConverter/commit/891f03100ba5dc4cfe83648d1949d8288ec7e218

**Pull request merged**